### PR TITLE
Internal-test-server

### PR DESCRIPTION
**Pull request by theganyo**
Enable use of an internal test server for unit testing.

### CODE REVIEW

**Comment by theganyo** [comment]
Opened this as a draft example of how we could update all of the unit tests to run without required an outside server to be started.

File: cmd/registry/cmd/upload/upload_test.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	"testing"
+
+	"github.com/apigee/registry/server/registry"
+	"github.com/apigee/registry/server/registry/test/grpctest"
+)
+
+// TestMain will set up a local RegistryServer and grpc.Server for all
+// tests in this package if APG_REGISTRY_ADDRESS env var is not set
+// for the client.
+func TestMain(m *testing.M) {
```
Comment (timburks) [inline cmd/registry/cmd/upload/upload_test.go:27]: I'm guessing that this would need to be done for every package under `cmd`, right?

**Review by timburks — COMMENTED** [review]
I don't know, this seems like a lot of magic to autostart a server that we might instead do in `make test` by just running `registry-server` right before running [`go test ./...`](https://github.com/apigee/registry/blob/257aefd033e819d318499092b9ef9d12148542ef/Makefile#L27) -- that call will fail if a server is already running (and we could catch and ignore the error there). This wouldn't require any code changes to the tests.

**Comment by theganyo** [comment]
Other than common expectation and ease of use, there are additional advantages to unit tests being self-contained. This way, each test is isolated from the others, so no potential side-effects, and testing can also be safely parallelized.

File: cmd/registry/cmd/upload/upload_test.go (hunk 1 above)
Comment (timburks) [reply cmd/registry/cmd/upload/upload_test.go:27]: That seems like a high line-count impact. Could this initialization be made implicit or moved up a level?

File: server/registry/test/grpctest/server.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	if err != nil {
+		gl.Close()
+		return nil, err
+	}
+
+	gl.Server = grpc.NewServer()
+	reflection.Register(gl.Server)
+	rpc.RegisterRegistryServer(gl.Server, rs)
+	rpc.RegisterAdminServer(gl.Server, rs)
+
```
Comment (timburks) [inline server/registry/test/grpctest/server.go:86]: This seems like a lot of code duplication.

**Comment by timburks** [comment]
@theganyo When we test the `registry-server` with the PostgreSQL backend, we just set up a PostgreSQL instance and run against that. I'm ok with that compromise (since we don't have a mock postgres to instantiate in our test) and I prefer the testing power of flexibility in this case as well, but I also am prepared to be outvoted (https://github.com/apigee/registry/issues/189) :).

File: server/registry/test/grpctest/server.go (hunk 2 above)
Comment (theganyo) [reply server/registry/test/grpctest/server.go:86]: Hmm. Do you mean with this? https://github.com/apigee/registry/blob/main/cmd/registry-server/main.go#L144. We could do a little refactoring once we decide this is a good direction.

File: cmd/registry/cmd/upload/upload_test.go (hunk 1 above)
Comment (theganyo) [reply cmd/registry/cmd/upload/upload_test.go:27]: I believe [TestMain()](https://pkg.go.dev/testing#hdr-Main) is the best / only official Go test mechanism to do this and it's package-specific.

**Comment by timburks** [comment]
https://github.com/apigee/registry/issues/611 describes another testing-related concern that's possibly competing with this one.

**Comment by theganyo** [comment]
> #611 describes another testing-related concern that's possibly competing with this one.

I'm not sure that competes, I believe unit tests should be a completely different concern than conformance testing.

File: server/registry/test/grpctest/server.go (hunk 2 above)
Comment (timburks) [reply server/registry/test/grpctest/server.go:86]: That's right. That's not in `server/registry` because I wanted to be able to mix in other servers (like a search server) from other future locations (at one time there was a search subdirectory that is now in registry-experimental). But a helper somewhere that does common work seems appropriate if you do this.

File: cmd/registry/cmd/upload/upload_test.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// Copyright 2020 Google LLC. All Rights Reserved.
```
Comment (seaneganx) [inline cmd/registry/cmd/upload/upload_test.go:1]: 2022

File: server/registry/test/grpctest/server.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
+package grpctest
```
Comment (seaneganx) [inline server/registry/test/grpctest/server.go:1]: This file should probably have a license.

File: server/registry/test/grpctest/server.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+// APG_REGISTRY_ADDRESS, APG_REGISTRY_AUDIENCES, and APG_REGISTRY_INSECURE
+// env vars are set for the client to connect to the created grpc service.
+// Call Close() when done to close server and clean up tmpDir as needed.
+// Example:
+// func TestXXX(t *testing.T) {
+// 	 l, err := grpctest.NewServer(registry.Config{})
+// 	 if err != nil {
+// 		t.Fatal(err)
+// 	 }
+// 	 defer l.Close()
```
[the remark this answers is not in the document]
Comment (theganyo) [reply server/registry/test/grpctest/server.go:53]: testing.T is unavailable here. This defer won't happen until the test is finished.

File: server/registry/test/grpctest/server.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+	"github.com/apigee/registry/server/registry"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/reflection"
+)
+
+// NewIfNoAddress will create a RegistryServer served by a
+// basic grpc.Server if env var APG_REGISTRY_ADDRESS is
+// not set, otherwise will simply return a dummy Server
+// assuming the tests will connect to a the remote address.
+func NewIfNoAddress(rc registry.Config) (*Server, error) {
```
[the remark this answers is not in the document]
Comment (theganyo) [reply server/registry/test/grpctest/server.go:21]: This function is exported for test functions that want to set up an instance per iteration.

**Comment by theganyo** [comment]
This change will isolate tests by giving them each their own DB. However, it's worth noting that because this preserves the existing client semantics that rely on global env vars for connection information, the tests cannot currently be parallelized whilst maintaining this isolation.

**Comment by theganyo** [comment]
This is ready for final review.

File: connection/grpctest/server.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+// 	 defer l.Close()
+//   ... run test here ...
+// }
+func NewIfNoAddress(rc registry.Config) (*Server, error) {
+	addr := os.Getenv("APG_REGISTRY_ADDRESS")
+	if addr != "" {
+		log.Printf("Client will use remote database at: %s", addr)
+		return nil, nil
+	}
+	log.Println("Client will use embedded SQLite3 database")
```
Comment (timburks) [inline connection/grpctest/server.go:48]: Is there a way (or is there no way) to configure the automatically-generated server to use a Postgres database?

Comment (timburks) [reply connection/grpctest/server.go:48]: I want to be able to detect (and rule-out) database-specific problems that might appear in the cmd tests.

Most database-specific problems should be found by the core server tests (in server/registry), and these can be run locally with postgres, which is great. If we assume that the core tests are complete, then we wouldn't need to run the cmd tests against a postgres database. But to me, that assumption takes an unnecessary risk for the arguable convenience of not having to start a test server.

Comment (theganyo) [reply connection/grpctest/server.go]: Aside: I'm just going to drop this link here in case we decide to run unit tests with an embedded postgres: https://github.com/fergusstrange/embedded-postgres.

Comment (timburks) [reply connection/grpctest/server.go]: I just created #624 to evaluate embedded postgres for unit and CI testing.

Summarizing our face-to-face discussion, we agreed to move forward with this PR with a new near-term goal to replace the use of environment variables to configure the registry client with configuration that can be read from a file. This file can be managed by the registry tool (allowing commands like `registry auth login` and `registry config set location`) and could point to arbitrary servers with arbitrary database backends. We think this is useful for allowing `registry config set project` and `registry config set location`, which could allow shorter relative resource names to be provided to other `registry` subcommands like `registry list`.

Most relevant to this PR, by removing the dependence on global values (environment variables) to configure tests, that reduces the uncertainty about what is happening that is expressed in all the comments like this one:
```
// TestMain will set up a local RegistryServer and grpc.Server for all
// tests in this package if APG_REGISTRY_ADDRESS env var is not set
// for the client.
```

tl;dr for now we're moving to a slightly more complex state that fixes a short-term problem and will make future improvements (issue TBD) to improve overall simplicity and clarity.

**Review by timburks — APPROVED** [review]
